### PR TITLE
Oracle relational operators with spaces & updates to MERGE INTO

### PR DESCRIPTION
**Pull request by davelindquist-egistix**
Includes two changes:

**Added ability to have operators like '>=' or '<=' separated by a space.  This includes:**

- Modifying the JJT syntax to support the 'space in the middle' versions of operators.
- Modifying the various operators to inherit from a new 'ComparisonOperator' class, which handles the (previously NotEqualsTo-only) logic for capturing the form of the operator.
- Giving each of the various operators a constructor that accepts the syntax used.
- Modifying TestUtils to strip comments out before comparing SQL text (necessary because condition07.sql is now passing, and has a comment).
- Updating SpecialOracleTest to indicate 130 tests passing now (condition7.sql now passes).
- Adding a new test specifically for operators into SpecialOracleTest.

**Corrected "MERGE INTO" parsing for more complicated statements.  Specifically:**

- Changed "Condition" to "Expression" for the "ON" clause -- this is needed to handle "ON" clauses that have "a = y AND b = z" or other more complicated expressions (basically the same as the "ON" clause in a SELECT query).
- Corrected the toString on the MergeUpdate clause, which was missing a comma between the fields.
- Added a new, more complicated MERGE INTO statement to the MergeTest class.

### CODE REVIEW

File: src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt [hunk 1]
```diff
@@ -1616,9 +1616,13 @@ Expression RegularCondition():
 	| "<" { result = new MinorThan(); }
 	| "=" { result = new EqualsTo(); }
 	| ">=" { result = new GreaterThanEquals(); }
+	| "> =" { result = new GreaterThanEquals("> ="); }
 	| "<=" { result = new MinorThanEquals(); }
+	| "< =" { result = new MinorThanEquals("< ="); }
 	| "<>" { result = new NotEqualsTo(); }
+	| "< >" { result = new NotEqualsTo("< >"); }
```
Comment (wumpz) [inline src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt]: Isnt this only half the way? Oracle allows, as I learned right now :), multiple whitespaces between < and =. Addionaly wouldnt it be better to map <..= to <=? Is there a case one need to reproduce oracles special space separated syntax?

Comment (davelindquist-egistix) [inline src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt]: Yes, you are correct that I was handling only a single space.  Ideally, I should handle any whitespace between it – any tips for me on what that would look like in JavaCC?  (I’m having difficulty reconciling the ‘skip’ characters with the insertion of variable whitespace in the token…)

I had originally started by mapping “< =” to just GreaterThanEquals(), but that unfortunately loses information about the ‘style’ of the operator (for example, I can’t go back and slap the hands of the developer for using non-standard relational operators).  I elected to follow the pattern of “NotEqualsTo”, which passed the actual operator text into the constructor.  Is that OK?  Did you have a preference for an alternate way of doing this?

Thanks!

**Comment by tejoe** [comment]
There are two errors here.
First oracle actually allows an arbitrary number of spaces in operators, so 
"<=" equals "< =" equals "<                 ="
you only considered a single space

Second and this one is much more remarkable "! =" in oracle is not the same like "! =" in postgresql. In Postgres "! =" is not an operator. It is two operators faculty and equals. 

**ORACLE**
select * from dual where 1 ! = 1
returns no row, because the condition evaluates to false

**POSTGRESQL**
select 1 where 1 ! = 1
returns a row, because the faculty of 1 equals 1

**Comment by davelindquist-egistix** [comment]
Hmm, that is interesting – I had not realized that postgres would actually interpret that as something.

Can you explain to me what the “! =” does in postgres?  (Or point me to the docs?)  I’m not quite clear on how it functions, and wasn’t able to turn anything up on a quick google.

In that case, it seems to me that we’d have to handle “! =” in the postgresql way?  (As two operators?)

Thanks!

**Comment by tejoe** [comment]
here is the operator docu for postgres:
http://www.postgresql.org/docs/9.5/static/typeconv-oper.html
example 10-1 will show you the "!" (factorial) operator

If there is an blank between ! and = postgres will always use the ! as factorial. I have no documentation for this, this is what I observed.

**Comment by davelindquist-egistix** [comment]
Thanks!  I think I turned up the same page a second before I got your reply. :-)

File: src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt (hunk 1 above)
Comment (davelindquist-egistix) [inline src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt]: I'll provide an updated commit to address:
a. The 'any amount of whitespace' issue (between parts of the operators)
b. Removal of "space between ! and =", as this means something different in PostgreSQL

**Comment by davelindquist-egistix** [comment]
OK, I've pushed another couple of commits -- this one allows for truly variable whitespace in the operators, and retracts the "! =" case, as it means something different in PostgreSQL  (I haven't tackled adding the "factorial" operator, which sounds fun.)

Feedback would be appreciated! :-)

**Comment by wumpz** [comment]
Again the question, why would one preserve the whitespaces? Isn't it for parsing purposes enough to map <  = to <= and therefore to less and equal?

Could you split your pull request into two. It is much easier to handle if I get one for one task or solution.

**Comment by davelindquist-egistix** [comment]
Sure, I'll split the pull request into two.

With regards to preserving the whitespace:
a. The test cases currently require preservation of the whitespace -- the parse/deparse test structure will fail if the whitespace is stripped.  If we elect to strip the whitespace, I imagine we'll need to modify the parse/deparse tests to do a regex-replace to remove whitespace?
b. The pre-existing NotEqualsTo structure established a pattern of preserving the operator (that is, it does not collapse "!=" into "<>"), so I followed that pattern.
c. For my particular usage, it is nice to be able to detect that a 'non-standard' form was used for the operator -- preserving the exact form of the token parsed as the operator was the easiest way to do so.

In order to prevent rework, could you give me an idea of what you'd like to see?  For example:

1. The existing structure, preserving the token that was parsed, OR
2. A modified structure that does not preserve the token that was parsed, but does provide a flag indicating that the operator was "non-standard", OR
3. Neither, just collapse the operators into standard form regardless of style used.

Additional Question: In either of cases 2 or 3, should we then collapse "!=" into "<>" as well?

(I don't have any super-strong preferences in either direction -- I'd prefer to have 1 or 2 over 3, but I can live with 3 if it is what you'd like.)

Thanks!
